presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java

```diff
@@ -224,8 +224,14 @@ static QueryStateMachine beginWithTicker(
                 ticker,
                 metadata,
                 warningCollector);
-        queryStateMachine.addStateChangeListener(newState -> QUERY_STATE_LOG.debug("Query %s is %s", queryStateMachine.getQueryId(), newState));
 
+        queryStateMachine.addStateChangeListener(newState -> {
+            QUERY_STATE_LOG.debug("Query %s is %s", queryStateMachine.getQueryId(), newState);
+            // mark finished or failed transaction as inactive
+            if (newState.isDone()) {
```

Comment:
@picrinite Did you check that this listener is triggered when the analysis fails (the case with permission error)? 

The last time we discussed it, we were thinking about wrapping the analysis into a try-catch, because we thought that the state change listener may not always be triggered. Is that not the case?

Reply:
@arhimondr The first commit is an improvement to make it safer for "setInactive" job. It is not triggered for permission error.
The second commit guarantees "setInactive" job is triggered when there is permission error, but there are some tests failed and I don't know how to reproduce. Help appreciated !
Regarding  automatic test,  cleanup of transaction manager is an asynchronous job with a delay (current default is 5 minutes),  it is not deterministic.

Reply:
An integration test is added,   the added test will fail without current fix.